src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs

```diff
@@ -0,0 +1,224 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ChangeNamespace;
+using Microsoft.CodeAnalysis.CodeRefactorings.MoveType;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.MoveToNamespace
+{
+    internal interface IMoveToNamespaceService : ILanguageService
+    {
+        Task<ImmutableArray<AbstractMoveToNamespaceCodeAction>> GetCodeActionsAsync(Document document, TextSpan span, CancellationToken cancellationToken);
+        Task<MoveToNamespaceAnalysisResult> AnalyzeTypeAtPositionAsync(Document document, int position, CancellationToken cancellationToken);
+        Task<MoveToNamespaceResult> MoveToNamespaceAsync(MoveToNamespaceAnalysisResult analysisResult, string targetNamespace, CancellationToken cancellationToken);
+        MoveToNamespaceOptionsResult GetChangeNamespaceOptions(Document document, string defaultNamespace, ImmutableArray<string> namespaces);
+    }
+
+    internal abstract class AbstractMoveToNamespaceService<TNamespaceDeclarationSyntax, TNamedTypeDeclarationSyntax>
+        : IMoveToNamespaceService
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TNamedTypeDeclarationSyntax : SyntaxNode
+
+    {
+        protected abstract string GetNamespaceName(TNamespaceDeclarationSyntax syntax);
+        protected abstract string GetNamespaceName(TNamedTypeDeclarationSyntax syntax);
+
+        public async Task<ImmutableArray<AbstractMoveToNamespaceCodeAction>> GetCodeActionsAsync(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var typeAnalysisResult = await AnalyzeTypeAtPositionAsync(document, span.Start, cancellationToken).ConfigureAwait(false);
+
+            if (typeAnalysisResult.CanPerform)
+            {
+                return ImmutableArray.Create(AbstractMoveToNamespaceCodeAction.Generate(this, typeAnalysisResult));
+            }
+
+            return ImmutableArray<AbstractMoveToNamespaceCodeAction>.Empty;
+        }
+
+        public async Task<MoveToNamespaceAnalysisResult> AnalyzeTypeAtPositionAsync(
+            Document document,
+            int position,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(position);
+            var node = token.Parent;
+
+            var symbolInfo = semanticModel.GetSymbolInfo(node, cancellationToken: cancellationToken);
+            var symbol = symbolInfo.Symbol;
+            string @namespace = null;
+
+            if (symbol is INamespaceSymbol namespaceSymbol && !(node is TNamespaceDeclarationSyntax))
```

Comment:
Why is the symbol check required? Is it possible to circumvent with weird caret positions?

Reply:
Removed in favor of bounds checking in the namespace decl from the position